pan_os_upgrade/upgrade.py

```diff
@@ -1539,116 +1501,148 @@ def perform_reboot(firewall: Firewall, ha_details: Optional[dict] = None) -> Non
-    - Firewall not ready for the intended upgrade.
-    - Critical issues that prevent script continuation.
-    - Successful completion of a dry run.
-    - HA peer state is not synchronized (for HA setups).
+    Returns
+    -------
+    dict
+        A dictionary representation of the XML element. The dictionary mirrors the structure of the XML,
+        with tags as keys and text content or nested dictionaries as values. Elements with the same tag
+        are aggregated into a list.
 
-    Example Usage:
-    ```bash
-    python upgrade.py --hostname 192.168.1.1 --username admin --password secret --version 10.2.7
-    ```
-    This command will start the upgrade process for the firewall at '192.168.1.1' to version '10.2.7'.
+    Notes
+    -----
+    - This function is designed to work with PAN-OS XML API responses, which often use the 'entry' tag
+      to denote list items.
+    - The function does not preserve attributes of XML elements; it focuses solely on tags and text content.
+
+    Example
+    -------
+    Converting an XML element with nested children to a dictionary:
+        >>> xml_string = "<root><child>value</child><child><subchild>subvalue</subchild></child></root>"
+        >>> xml_element = ET.fromstring(xml_string)
+        >>> flatten_xml_to_dict(xml_element)
+        {'child': ['value', {'subchild': 'subvalue'}]}
     """
+    result = {}
+    for child_element in element:
+        child_tag = child_element.tag
 
-    # Create necessary directories
-    directories = [
-        "logs",
-        "assurance",
-        "assurance/configurations",
-        "assurance/readiness_checks",
-        "assurance/reports",
-        "assurance/snapshots",
-    ]
-    for dir in directories:
-        ensure_directory_exists(os.path.join(dir, "dummy_file"))
+        if child_element.text and len(child_element) == 0:
+            result[child_tag] = child_element.text
+        else:
+            if child_tag in result:
+                if type(result.get(child_tag)) is not list:
+                    result[child_tag] = [
+                        result.get(child_tag),
+                        flatten_xml_to_dict(child_element),
+                    ]
+                else:
+                    result[child_tag].append(flatten_xml_to_dict(child_element))
+            else:
+                if child_tag == "entry":
+                    # Always assume entries are a list.
+                    result[child_tag] = [flatten_xml_to_dict(child_element)]
+                else:
+                    result[child_tag] = flatten_xml_to_dict(child_element)
 
-    # Configure logging right after directory setup
-    configure_logging(log_level)
+    return result
 
-    # Create our connection to the firewall
-    logging.debug(f"{get_emoji('start')} Connecting to PAN-OS firewall...")
-    firewall = connect_to_firewall(
-        hostname=hostname,
-        api_username=username,
-        api_password=password,
+
+def model_from_api_response(
```

Comment:
What is the expected return object here?

Reply:
Added annotation for this. it returns any model that implements `FromAPIResponseMixin`